From 60f7d8dd5e93caa40636c6ecabb8e608f7a1bdf5 Mon Sep 17 00:00:00 2001
From: Povilas Staniulis <wdmonster@gmail.com>
Date: Sun, 28 May 2017 23:10:59 +0300
Subject: [PATCH] HACK: netd: disable output interface result validation

The cause of these issues appears to be Android's netd daemon not
playing well with the latest kernel. I wasn't able to find the root
cause, but I did hack netd to enable networking to work.

WiFi now works fine on my T100TA

Conflicts:
	server/RouteController.cpp

Change-Id: Ibba952588d3a3934265c297d0f115584c23d9ecd
---
 server/RouteController.cpp | 18 +++++++++++++++---
 1 file changed, 15 insertions(+), 3 deletions(-)

diff --git a/server/RouteController.cpp b/server/RouteController.cpp
index c78854d8..6b702acf 100644
--- a/server/RouteController.cpp
+++ b/server/RouteController.cpp
@@ -723,17 +723,29 @@ WARN_UNUSED_RESULT int RouteController::modifyPhysicalNetwork(unsigned netId, co
     }
 
     if (int ret = modifyIncomingPacketMark(netId, interface, permission, add)) {
+        ALOGE("DDS: modifyIncomingPacketMark failed with error: %s", strerror(-ret));
         return ret;
     }
     if (int ret = modifyExplicitNetworkRule(netId, table, permission, INVALID_UID, INVALID_UID,
                                             add)) {
-        return ret;
+        ALOGE("DDS: modifyExplicitNetworkRule failed with error: %s", strerror(-ret));
+	return ret;
     }
-    if (int ret = modifyOutputInterfaceRules(interface, table, permission, INVALID_UID, INVALID_UID,
+   /* DDS: HACK ! This can sometimes fail on latest kernels, leading to a non-working network. Disable validation for now, until we find a better fix. */
+   if (int ret = modifyOutputInterfaceRules(interface, table, permission, INVALID_UID, INVALID_UID,
                                             add)) {
-        return ret;
+        ALOGE("DDS: modifyOutputInterfaceRules failed with error: %s, ignoring this.", strerror(-ret));
+    //  return ret;
     }
 
+   int ret2 = modifyImplicitNetworkRule(netId, table, permission, add);
+
+   if (ret2) {
+      ALOGE("DDS: modifyImplicitNetworkRule failed with error: %s", strerror(-ret2));
+   }
+
+   return ret2;
+
     // Only set implicit rules for networks that don't require permissions.
     //
     // This is so that if the default network ceases to be the default network and then switches
-- 
2.17.1

